funsor/contract.py

```diff
@@ -0,0 +1,242 @@
+from __future__ import absolute_import, division, print_function
+
+from collections import OrderedDict
+from six import integer_types
+
+import opt_einsum
+import torch
+
+import funsor.ops as ops
+from funsor.distributions import Gaussian, Delta
+from funsor.interpreter import interpretation, reinterpret
+from funsor.optimizer import Finitary, optimize
+from funsor.sum_product import _partition
+from funsor.terms import Funsor, Number, Reduce, Variable, eager
+from funsor.torch import Tensor
+
+
+ATOMS = (Tensor, Gaussian, Delta, Number, Variable)
+
+
+def _make_base_lhs(arg, reduced_vars, normalized=False):
```

Comment:
This should no longer be used, right?

Reply:
Moved to `einsum` since I'm only using it for testing